### ISSUE TITLE
Why is rand distance treating eigenvalues as if they are probability densities?

### ISSUE THREAD

**Issue by jogardi** [CalculatedContent/WeightWatcher#104]
On line 1306 of weightwatcher.py the sorted eigenvalues are plugged into the jensen shannon divergence just as if the eigenvalues were counts in a histogram. I understand that this still is a divergence in that the number will be bigger for a less similar set of eigenvalues. And you have shown that this works imperically. But this isn't making theoretical sense to me.  When you plot the ESD you use bins as I would expect. This may be an ambiguity in documentation and not a code issue.

**Comment by charlesmartin14** [comment]
Thanks for pointing this out
It's a very new idea and has not totally been flushed out yet

I'm happy to try other things that work more consistently and are better theoretically justified
For example, we  could try to estimate the density of the EDD using a KDE, which is what is done
when running the MP fits

**Comment by jogardi** [comment]
I think using doing a histogram the same way as you do for plotting the ESD should work. Except you would need the bins to be the same each time you make the histogram rather than dynamicly sizing the bins like matplotlib does under the hood.

By the way, do you think that the heavy tailed distribution fits with the manifold and information bottleneck interpretations of how deep nets generalize? You show heavy tailed matrices are often lower rank. That is equivalent to saying they are projecting onto a lower dimensional manifold or that they are throwing away information. The idea is that the information being lost must be the noise since it is the result of training. 

I know there are issues with the manifold and information bottleneck interpretations:
- the good generalization with over-parameterization would seem to contradict those interpretations
- It makes people worry about tuning the dimensionality of their manifold like researchers in autoencoders tune the embedding dimension
- the loss surface might not be convex if you use minimal dimension. So there was room for improvement in those ideas.

**Comment by charlesmartin14** [comment]
1) In quantum physics, the generalized matrix entropy is defined directly in terms of the eigenvalues.  So its not clear to me why we can't just do this with an arbitrary cross-entropy /  divergence as well.  I'm open to trying other things if they work well from an engineering point of view, but, theoretically, I don't see the problem

2) The heavy-tailed weight matrices *do not* lose rank.  In fact, we specifically show this in our JMLR paper
The do lose soft-rank, but that's different

3) However, we do find that we can get an estimate of the out-of-sample test accuracy by evaluating the training accuracy on a low-rank approximation of each layer.  WeightWatcher supports this in the SVDSmoothing function. And this seems to work in some cases, but not all.  We don't know why yet.

4) I dont know what to say about the manifold and information bottleneck interpretations:

**Comment by jogardi** [comment]
Thanks. Now I see that you discussed information bottleneck in your jlmr paper.  

This is also like the way jpeg is denoising. The quantization coefficients are analogous to the eigenvalues, and the DCT transform is like the eigenvectors. The smaller quantization coefficients correspond to the high frequency component from the DCT because that is the noise. So JPEG is also losing soft rank rather than hard rank. 

If the smaller eigenvalues correspond to overfitting then that that would explain why SVD smoothing gives the test accuracy.

**Comment by charlesmartin14** [comment]
< If the smaller eigenvalues correspond to overfitting then that that would explain why SVD smoothing gives the test accuracy.


in our most recent paper, we motivate SVDSmoothing is a data-dependent shape metric
